File: frontend/bff/src/infrastructure/grpc/mappers/counterparty.contract-mapper.ts
import type { CreateCounterpartyCommand } from '../../../application/counterparties/commands/create-counterparty.command.js';
import type { CounterpartyDto } from '../../../application/counterparties/dtos/counterparty.dto.js';
import { nullable, nullableNumber } from './common.contract-mapper.js';
import { counterpartyTypeEnum } from './enum.contract-mapper.js';

export interface ContractCounterparty {
  id: string;
  name: string;
  type: string;
  document?: string;
  address?: string;
  country?: string;
  isHomologated: boolean;
  notionalLimitUsd?: number;
  mtmLimitUsd?: number;
}

export const toCounterpartyDto = (c: ContractCounterparty): CounterpartyDto => ({
  id: c.id,
  name: c.name,
  type: counterpartyTypeEnum.fromContractRequired(c.type),
  document: nullable(c.document),
  address: nullable(c.address),
  country: nullable(c.country),
  isHomologated: c.isHomologated,
  notionalLimitUsd: nullableNumber(c.notionalLimitUsd),
  mtmLimitUsd: nullableNumber(c.mtmLimitUsd),
});

export const toContractCounterpartyInput = (c: Omit<CreateCounterpartyCommand, 'context'>) => ({
  name: c.name,
  type: counterpartyTypeEnum.toContract(c.type),
  document: c.document,
  address: c.address,
  country: c.country,
  notionalLimitUsd: c.notionalLimitUsd,
  mtmLimitUsd: c.mtmLimitUsd,
});

File: backend/core-service/src/Fix.Presentation/Mappers/CounterpartyContractMapper.cs
using Fix.Application.Counterparties.Dtos;
using Contract = Fix.Contracts.V1;

namespace Fix.Presentation.Mappers;

internal static class CounterpartyContractMapper
{
    public static Contract.Counterparty ToContract(this CounterpartyDto counterparty)
    {
        var contract = new Contract.Counterparty
        {
            Id = counterparty.Id.ToString(),
            Code = counterparty.Code,
            Name = counterparty.Name,
            Type = counterparty.Type.ToContract<Contract.CounterpartyType>(),
            IsHomologated = counterparty.IsHomologated,
        };

        if (counterparty.Document is { } document) contract.Document = document;
        if (counterparty.Address is { } address) contract.Address = address;
        if (counterparty.Country is { } country) contract.Country = country;
        if (counterparty.NotionalLimitUsd is { } notional) contract.NotionalLimitUsd = (double)notional;
        if (counterparty.MtmLimitUsd is { } mtm) contract.MtmLimitUsd = (double)mtm;
        return contract;
    }
}
